### PR TITLE
Use PATCH request to update EventMesh subscription WebhookAuth config

### PR DESCRIPTION
**Pull request by marcobebway**
**Description**

Use PATCH request to update EventMesh subscription WebhookAuth config.

**Related issue(s)**

- https://github.com/kyma-project/kyma/issues/17370.

### CODE REVIEW

**Review by NHingerl — DISMISSED** [review]

**Review by NHingerl — APPROVED** [review]

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go [hunk 1]
```diff
@@ -146,6 +146,13 @@ func (em *EventMesh) SyncSubscription(subscription *eventingv1alpha2.Subscriptio
 			log.Errorw("Failed to handle EventMesh subscription modified", errorLogKey, err)
 			return false, err
 		}
+
+		if featureflags.IsEventingWebhookAuthEnabled() && !isEventMeshSubModified {
```
Comment (mfaizanse) [inline components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:150]: (nit) Can you add a comment explaining why we check for `!isEventMeshSubModified`?

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go [hunk 2]
```diff
@@ -269,6 +287,67 @@ func (em *EventMesh) handleCreateEventMeshSub(eventMeshSub *types.Subscription,
...
+	state := types.State{Action: types.StateActionPause}
+	resp, err := em.client.UpdateState(eventMeshSub.Name, state)
+	if err != nil {
+		return fmt.Errorf("failed to pause EventMesh subscription: %w", err)
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		em.namedLogger().Warnf("failed to pause EventMesh subscription: %s, not found", eventMeshSub.Name)
+		return nil
+	}
+	if resp.StatusCode != http.StatusAccepted {
```
Comment (mfaizanse) [inline components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:314]: If subscription is already paused then will the request fail or will it still return `http.StatusAccepted`?

File: components/eventing-controller/testing/eventmeshmock.go [hunk 3]
```diff
@@ -202,6 +229,53 @@ func GetSubscriptionResponse(m *EventMeshMock) ResponseWithName {
...
+				{
+					panic(fmt.Sprintf("EventMesh subscription status is not supported: %#v", state))
+				}
+			}
+
+			m.Subscriptions.PutSubscription(key, subscription)
+			w.WriteHeader(http.StatusAccepted)
+
+			err := json.NewEncoder(w).Encode(*subscription)
+			Expect(err).ShouldNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/testing/eventmeshmock.go:271]: No need to return an error. The `Expect(err).ShouldNot(HaveOccurred())` statement will fail the test.

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go [hunk 4]
```diff
@@ -269,6 +287,67 @@ func (em *EventMesh) handleCreateEventMeshSub(eventMeshSub *types.Subscription,
...
+	state = types.State{Action: types.StateActionResume}
+	resp, err = em.client.UpdateState(eventMeshSub.Name, state)
+	if err != nil {
+		return fmt.Errorf("failed to resume EventMesh subscription: %w", err)
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		em.namedLogger().Warnf("failed to resume EventMesh subscription: %s, not found", eventMeshSub.Name)
+		return nil
+	}
+	if resp.StatusCode != http.StatusAccepted {
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:343]: It returns http.StatusAccepted.

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go (hunk 1 above)
Comment (marcobebway) [reply components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:152]: Done

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go [hunk 5]
```diff
@@ -218,13 +225,24 @@ func (em *EventMesh) getProcessedEventTypes(kymaSubscription *eventingv1alpha2.S
...
+		hash := kymaSub.Status.Backend.EventMeshLocalHash
+		if isModified, err = backendutils.IsEventMeshSubModified(eventMeshSub, hash); err != nil {
+			return false, err
+		}
+	} else {
+		hash := kymaSub.Status.Backend.Ev2hash
+		if isModified, err = backendutils.IsEventMeshSubModified(eventMeshSub, hash); err != nil {
+			return false, err
+		}
 	}
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:244]: Done

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go [hunk 6]
```diff
@@ -269,6 +287,67 @@ func (em *EventMesh) handleCreateEventMeshSub(eventMeshSub *types.Subscription,
 	return eventMeshServerSub, nil
 }
 
+// handleWebhookAuthChange handles the EventMesh subscription WebhookAuth changes.
+// It uses the PATCH request API provided by EventMesh to update the subscription WebhookAuth.
+func (em *EventMesh) handleWebhookAuthChange(eventMeshSub *types.Subscription,
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:293]: I will add unit tests in a follow-up PR.

For the time being, it is tested on integration level here:

- https://github.com/marcobebway/kyma/tree/17370-use-patch-request-to-change-webhook-credentials/components/eventing-controller/controllers/subscription/eventmesh/test
- https://github.com/marcobebway/kyma/tree/17370-use-patch-request-to-change-webhook-credentials/components/eventing-controller/controllers/subscription/eventmesh/testwithory
- https://github.com/marcobebway/kyma/tree/17370-use-patch-request-to-change-webhook-credentials/components/eventing-controller/controllers/subscription/eventmesh/testwebhookauth

**Review by NHingerl — APPROVED** [review]

**Review by mfaizanse — APPROVED** [review]

**Pull request merged**